File: app/android/src/main/kotlin/io/gratia/app/sensors/MagnetometerManager.kt
package io.gratia.app.sensors

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener as AndroidSensorEventListener
import android.hardware.SensorManager
import android.util.Log
import kotlin.math.sqrt

/**
 * Magnetometer (magnetic field sensor) manager (optional).
 *
 * Reads geomagnetic field data from TYPE_MAGNETIC_FIELD at low frequency.
 * This is an optional sensor that contributes +4 to the Composite Presence
 * Score but is NOT required for the PoL threshold.
 *
 * Magnetometer data serves two purposes:
 * 1. Environmental fingerprinting — indoor magnetic fields vary by location,
 *    providing weak but useful location corroboration.
 * 2. Oracle layer (Phase 3) — aggregated magnetic field data can detect
 *    geomagnetic anomalies.
 *
 * WHY optional: While most post-2015 smartphones include a magnetometer,
 * some budget devices omit it. The PoL threshold cannot depend on it.
 */
class MagnetometerManager(
    private val context: Context
) {

    companion object {
        private const val TAG = "GratiaMagnetometer"

        // WHY: 5-minute sampling is sufficient. Earth's magnetic field is
        // quasi-static, and indoor magnetic anomalies change only when the
        // phone moves to a different location.
        private const val SAMPLE_INTERVAL_MS = 5L * 60 * 1000 // 5 minutes
    }

    private val sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as? SensorManager
    private val magnetometer = sensorManager?.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD)

    private var isRunning = false
    private var lastSampleTimeMs = 0L

    /** Most recent magnetic field magnitude in microtesla. */
    @Volatile
    var currentMagnitudeMicroTesla: Float = 0f
        private set

    /** Most recent raw X/Y/Z components in microtesla. */
    @Volatile
    var currentX: Float = 0f
        private set

    @Volatile
    var currentY: Float = 0f
        private set

    @Volatile
    var currentZ: Float = 0f
        private set

    /** Whether at least one valid reading has been obtained. */
    @Volatile
    var hasReading: Boolean = false
        private set

    /** Optional callback for magnetic field updates. */
    var onMagneticFieldUpdate: ((magnitude: Float, x: Float, y: Float, z: Float) -> Unit)? = null

    private val sensorListener = object : AndroidSensorEventListener {
        override fun onSensorChanged(event: SensorEvent) {
            if (event.sensor.type != Sensor.TYPE_MAGNETIC_FIELD) return

            val now = System.currentTimeMillis()
            if (now - lastSampleTimeMs < SAMPLE_INTERVAL_MS && hasReading) return

            lastSampleTimeMs = now
            currentX = event.values[0]
            currentY = event.values[1]
            currentZ = event.values[2]
            currentMagnitudeMicroTesla = sqrt(
                currentX * currentX + currentY * currentY + currentZ * currentZ
            )
            hasReading = true

            Log.d(TAG, "Magnetic field: magnitude=${currentMagnitudeMicroTesla}uT " +
                "(x=$currentX, y=$currentY, z=$currentZ)")
            onMagneticFieldUpdate?.invoke(currentMagnitudeMicroTesla, currentX, currentY, currentZ)
        }

        override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {
            when (accuracy) {
                SensorManager.SENSOR_STATUS_UNRELIABLE ->
                    Log.d(TAG, "Magnetometer accuracy: UNRELIABLE (needs calibration)")
                SensorManager.SENSOR_STATUS_ACCURACY_LOW ->
                    Log.d(TAG, "Magnetometer accuracy: LOW")
                SensorManager.SENSOR_STATUS_ACCURACY_MEDIUM ->
                    Log.d(TAG, "Magnetometer accuracy: MEDIUM")
                SensorManager.SENSOR_STATUS_ACCURACY_HIGH ->
                    Log.d(TAG, "Magnetometer accuracy: HIGH")
            }
        }
    }

    /**
     * Start reading magnetic field data.
     *
     * Gracefully returns without error if the sensor is not available.
     */
    fun start() {
        if (isRunning) return

        if (sensorManager == null || magnetometer == null) {
            Log.d(TAG, "Magnetometer not available on this device (optional sensor)")
            return
        }

        sensorManager.registerListener(
            sensorListener,
            magnetometer,
            SensorManager.SENSOR_DELAY_NORMAL
        )

        isRunning = true
        Log.i(TAG, "Magnetometer tracking started")
    }

    /** Stop reading and release sensor resources. */
    fun stop() {
        if (!isRunning) return

        sensorManager?.unregisterListener(sensorListener)
        isRunning = false
        hasReading = false
        Log.i(TAG, "Magnetometer tracking stopped")
    }

    /** Check whether this manager is actively collecting data. */
    fun isActive(): Boolean = isRunning

    /** Check whether the magnetometer is present on this device. */
    fun isAvailable(): Boolean = magnetometer != null
}

File: app/ios/Gratia/Sources/Sensors/MagnetometerManager.swift
import Foundation
import CoreMotion
import CoreLocation

// ============================================================================
// MagnetometerManager — Core Motion magnetometer + compass heading
//
// Mirrors Android's MagnetometerManager.kt:
// - CMMotionManager magnetometer data
// - CLLocationManager heading (compass) for presence score
//
// WHY: Magnetometer data contributes +4 to the Composite Presence Score.
// It provides additional environmental context that is difficult to
// emulate accurately (requires real magnetic field readings).
//
// The magnetometer is an OPTIONAL sensor -- not required for threshold.
// ============================================================================

final class MagnetometerManager: NSObject, ObservableObject, CLLocationManagerDelegate {

    private let logger = GratiaLogger(tag: "GratiaMagnetometer")

    private let motionManager = CMMotionManager()
    private let locationManager = CLLocationManager()
    private var isRunning = false

    /// Most recent magnetic field strength (microteslas).
    @Published private(set) var magneticFieldStrength: Double = 0

    /// Most recent compass heading (degrees from true north).
    @Published private(set) var heading: Double = 0

    /// Whether the magnetometer is available on this device.
    var isAvailable: Bool {
        motionManager.isMagnetometerAvailable
    }

    override init() {
        super.init()
        locationManager.delegate = self
    }

    /// Start collecting magnetometer data.
    func start() {
        guard !isRunning else { return }

        guard isAvailable else {
            logger.warning("Magnetometer not available on this device")
            return
        }

        // WHY: 1-second update interval is sufficient for presence score.
        // Magnetometer data doesn't change rapidly in normal use.
        motionManager.magnetometerUpdateInterval = 1.0

        motionManager.startMagnetometerUpdates(to: .main) { [weak self] data, error in
            guard let self = self, let data = data else { return }

            let field = data.magneticField
            self.magneticFieldStrength = sqrt(
                field.x * field.x + field.y * field.y + field.z * field.z
            )
        }

        // WHY: Compass heading provides a human-readable orientation metric
        // and verifies the device has real magnetic sensors.
        if CLLocationManager.headingAvailable() {
            locationManager.startUpdatingHeading()
        }

        isRunning = true
        logger.info("Magnetometer tracking started")
    }

    /// Stop magnetometer data collection.
    func stop() {
        guard isRunning else { return }

        motionManager.stopMagnetometerUpdates()
        locationManager.stopUpdatingHeading()
        isRunning = false
        logger.info("Magnetometer tracking stopped")
    }

    /// Whether this manager is actively collecting data.
    var isActive: Bool { isRunning }

    // MARK: - CLLocationManagerDelegate (heading)

    func locationManager(_ manager: CLLocationManager, didUpdateHeading newHeading: CLHeading) {
        // WHY: trueHeading is relative to true north (requires location fix).
        // magneticHeading works without GPS. We prefer true heading when available.
        if newHeading.trueHeading >= 0 {
            heading = newHeading.trueHeading
        } else {
            heading = newHeading.magneticHeading
        }
    }
}
